Added case for a single location

Stormbringer68 opened the pull request:
Broke up the functionality in location_bounds to a case for N > 1 and a case for N ==1.  N > 1 is the original code (subtracting the case for N == 1).  For the N == 1 case, both the lower and upper bounds would be the same as the location (longitude) passed.

pbugnion commented:
This looks great. Thanks a lot for submitting it! I'll review it tomorrow.

pbugnion commented:
I think this is great. At the moment, it returns `lower_bound = upper_bound = longitudes[0]`, which means the gap between `lower_bound` and `upper_bound` is 0. This causes the `merge_longitude_bounds` function to fail (it assumes that lower_bound is always less than upper bound, up to a normalisation constant). Note that this problem predates your fix -- it was just masked by it.

It would seem reasonable to expect a `longitude_bounds` function to always return bounds that are strictly ordered (i.e. `lower_bound < upper_bound`). One way to do this would be:

```python
def longitude_bounds(longitudes):
    N = float(len(longitudes))
    if N > 1:
        # code for N > 1
    else:
        lower_bound = upper_bound = longitudes[0]
    lower_bound, upper_bound = _ensure_gap_in_bounds(lower_bound, upper_bound)
    return lower_bound, upper_bound

def _ensure_gap_in_bounds(lower_bound, upper_bound):
    if abs(upper_bound - lower_bound) < 2.0*EPSILON:
        lower_bound -= EPSILON
        upper_bound += EPSILON
    return lower_bound, upper_bound
```

For consistency, we could do the same in the `latitude_bounds` function.

Besides this, I have two suggestions:

 - Do you think it's more readable to have the special case first:

```python
if N == 1:
    lower_bound = upper_bound = longitudes[0]
else:
    # rest 
return lower_bound, upper_bound
```

- Should we duplicate the unit test for latitude bounds?

Thanks again for submitting this!

Stormbringer68 commented:
Hi Pascal,

I love your idea about adding and subtracting EPSILON to make proper lower
and upper bounds and think that is a better way to solve the problem and
handle the latitude bounds the same way.  On the two suggestions, I put the
regular case first as almost everyone will use that but the special case is
so small, I think it is better your way.  Also, I think having a single
point unit test for the latitude bounds is the appropriate thing to do.

I haven't altered a pull request before so I am not sure what is the proper
procedure.  Do you reject it and I do another or I just update the branch
on my repository and do another pull request?

Thanks,

--Chris

On Wed, Jul 12, 2017 at 2:20 AM, Pascal Bugnion <notifications@github.com>
wrote:

> I think this is great. At the moment, it returns lower_bound =
> upper_bound = longitudes[0], which means the gap between lower_bound and
> upper_bound is 0. This causes the merge_longitude_bounds function to fail
> (it assumes that lower_bound is always less than upper bound, up to a
> normalisation constant). Note that this problem predates your fix -- it was
> just masked by it.
>
> It would seem reasonable to expect a longitude_bounds function to always
> return bounds that are strictly ordered (i.e. lower_bound < upper_bound).
> One way to do this would be:
>
> def longitude_bounds(longitudes):
>     N = float(len(longitudes))
>     if N > 1:
>         # code for N > 1
>     else:
>         lower_bound = upper_bound = longitudes[0]
>     lower_bound, upper_bound = _ensure_gap_in_bounds(lower_bound, upper_bound)
>     return lower_bound, upper_bound
> def _ensure_gap_in_bounds(lower_bound, upper_bound):
>     if abs(upper_bound - lower_bound) < 2.0*EPSILON:
>         lower_bound -= EPSILON
>         upper_bound += EPSILON
>     return lower_bound, upper_bound
>
> For consistency, we could do the same in the latitude_bounds function.
>
> Besides this, I have two suggestions:
>
>    - Do you think it's more readable to have the special case first:
>
> if N == 1:
>     lower_bound = upper_bound = longitudes[0]else:
>     # rest return lower_bound, upper_bound
>
>
>    - Should we duplicate the unit test for latitude bounds?
>
> Thanks again for submitting this!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pbugnion/gmaps/pull/160#issuecomment-314665911>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI7ZQXIanvabzX0e-KaSZvy5KjHAqFjJks5sNGWwgaJpZM4OT16u>
> .
>

pbugnion commented:
That's great!

> I haven't altered a pull request before so I am not sure what is the proper
procedure.

You just push to your `single-point-longitude-bounds` branch on GitHub. GitHub will automatically add the commits to this PR. Another way to think of this is that PRs point to a branch (at least until that branch is merged) rather than to a specific commit.

If you've named your branch on GitHub 'origin', just:

```
git push origin single-point-longitude-bounds
```

Stormbringer68 commented:
Hi Pascal,

I updated the code.  I changed it slightly from your suggestion (but still
did your suggestion).  Since the original fix was lower_bound = upper_bound
= locations[0], by doing the ensure_gap_in_bounds, it just makes sure that
first assignment winds up not being used.  So, I put that EPSILON addition
/ subtraction in the N == 1 case directly and then I used your idea for
abs(upper_bound - lower_bound) < 2.0 * EPSILON for the unit tests (for both
latitude and longitude).  I hope that is OK.  I loved your idea and used
them both, I just didn't want to put the original solution line as it was a
throwaway.  I guess it is bad habit from learning on a TI 99 4/A with 16K.
 :-)

Thanks,

--Chris

On Fri, Jul 14, 2017 at 1:09 AM, Pascal Bugnion <notifications@github.com>
wrote:

> That's great!
>
> I haven't altered a pull request before so I am not sure what is the proper
> procedure.
>
> You just push to your single-point-longitude-bounds branch on GitHub.
> GitHub will automatically add the commits to this PR. Another way to think
> of this is that PRs point to a branch (at least until that branch is
> merged) rather than to a specific commit.
>
> If you've named your branch on GitHub 'origin', just:
>
> git push origin single-point-longitude-bounds
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pbugnion/gmaps/pull/160#issuecomment-315271351>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI7ZQcCo-7EPO7kFbTYBcuJUbsOYN-uzks5sNvgZgaJpZM4OT16u>
> .
>

pbugnion commented:
That looks great! Not binding to a temporary variable unnecessarily makes sense. My last minor comment is that I think it would be better to write:

```python
def longitude_bounds(longitudes):
    if len(longitudes) == 1:
        # code for N = 1
    else:
        N = float(len(longitudes))
        # rest of the code
    return lower_bound, upper_bound
```

i.e. switch from comparing `N` (which is a float) to comparing `len(longitudes)` (which is an int). Float equality makes me uneasy, even though it's well defined here. What do you think?

Thanks again for your hard work.

Stormbringer68 commented:
Hi Pascal,

Done and I did the same for latitude bounds (eliminating the float equality
test for the special cases).

Thanks,

--Chris

On Fri, Jul 14, 2017 at 4:43 AM, Pascal Bugnion <notifications@github.com>
wrote:

> That looks great! Not binding to a temporary variable unnecessarily makes
> sense. My last minor comment is that I think it would be better to write:
>
> def longitude_bounds(longitudes):
>     if len(longitudes) == 1:
>         # code for N = 1
>     else:
>         N = float(len(longitudes))
>         # rest of the code
>     return lower_bound, upper_bound
>
> i.e. switch from comparing N (which is a float) to comparing
> len(longitudes) (which is an int). Float equality makes me uneasy, even
> though it's well defined here. What do you think?
>
> Thanks again for your hard work.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pbugnion/gmaps/pull/160#issuecomment-315305899>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI7ZQUH53q_08-rbeKiferws7aARt3Yxks5sNyocgaJpZM4OT16u>
> .
>

pbugnion commented:
This is awesome. Thanks very much for this!

pbugnion commented:
This is now available in gmaps 0.5.4, which can already be installed via pip. It'll be available in conda-forge in a few hours.
